core-bundle/src/Security/Authenticator/ContaoLoginAuthenticator.php

```diff
@@ -0,0 +1,232 @@
+use Symfony\Component\Security\Core\User\UserProviderInterface;
+use Symfony\Component\Security\Http\Authentication\AuthenticationFailureHandlerInterface;
+use Symfony\Component\Security\Http\Authentication\AuthenticationSuccessHandlerInterface;
+use Symfony\Component\Security\Http\Authenticator\AbstractAuthenticator;
+use Symfony\Component\Security\Http\Authenticator\InteractiveAuthenticatorInterface;
+use Symfony\Component\Security\Http\Authenticator\Passport\Badge\CsrfTokenBadge;
+use Symfony\Component\Security\Http\Authenticator\Passport\Badge\PasswordUpgradeBadge;
+use Symfony\Component\Security\Http\Authenticator\Passport\Badge\RememberMeBadge;
+use Symfony\Component\Security\Http\Authenticator\Passport\Badge\UserBadge;
+use Symfony\Component\Security\Http\Authenticator\Passport\Credentials\PasswordCredentials;
+use Symfony\Component\Security\Http\Authenticator\Passport\Passport;
+use Symfony\Component\Security\Http\EntryPoint\AuthenticationEntryPointInterface;
+use Symfony\Component\Security\Http\ParameterBagUtils;
+
+class ContaoLoginAuthenticator extends AbstractAuthenticator implements AuthenticationEntryPointInterface, InteractiveAuthenticatorInterface
+{
+    private array $options;
+
+    public function __construct(
+        private UserProviderInterface $userProvider,
+        private AuthenticationSuccessHandlerInterface $successHandler,
+        private AuthenticationFailureHandlerInterface $failureHandler,
+        private ScopeMatcher $scopeMatcher,
+        private RouterInterface $router,
+        private UriSigner $uriSigner,
+        private ContaoFramework $framework,
+        private TokenStorageInterface $tokenStorage,
+        private PageRegistry $pageRegistry,
+        private HttpKernelInterface $httpKernel,
+        private TwoFactorAuthenticator $twoFactorAuthenticator,
+        array $options,
+    ) {
+        $this->options = array_merge([
+            'username_parameter' => 'username',
+            'password_parameter' => 'password',
+            'check_path' => '/login_check',
+            'post_only' => true,
+            'enable_csrf' => false,
+            'csrf_parameter' => '_csrf_token',
+            'csrf_token_id' => 'authenticate',
+        ], $options);
+    }
+
+    public function start(Request $request, AuthenticationException $authException = null): RedirectResponse|Response
+    {
+        if ($this->scopeMatcher->isBackendRequest($request)) {
+            return $this->redirectToBackend($request);
+        }
+
+        $this->framework->initialize();
+
+        $page = $request->attributes->get('pageModel');
+
+        if (!$page instanceof PageModel) {
+            throw new \RuntimeException('No PageModel found on request.');
+        }
+
+        $page->loadDetails();
+        $page->protected = false;
+
+        $route = $this->pageRegistry->getRoute($page);
+        $subRequest = $request->duplicate(null, null, $route->getDefaults());
```

Comment:
Looks like this is all unused if the token storage does not have a token?

Reply:
Yes, I tried to avoid the `else` clause.